File: Leetcode-solution-python/successfulPairsOfSpellsAndPotions.py
from typing import List


def lower_bound(arr: List[int], val: int):
    start, end = 0, len(arr) - 1
    while start <= end:
        mid = start + (end - start) // 2
        if arr[mid] > val:
            end = mid - 1
        elif arr[mid] < val:
            start = mid + 1
        else:
            end = mid - 1
    return start


class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        n, m = len(spells), len(potions)
        ans = [0] * n
        potions.sort()
        for i in range(n):
            req = success // spells[i]
            if success % spells[i] > 0:
                req += 1
            if req > potions[m - 1]:
                continue
            total_pairs = m - lower_bound(potions, req)
            ans[i] = total_pairs
        return ans

File: Leetcode-solution-c-c++/successfulPairsOfSpellsAndPotions.cpp
#include <bits/stdc++.h>

#define printArr(arr)                    \
    for (int i = 0; i < arr.size(); i++) \
        cout << arr[i] << " ";           \
    cout << endl;
#define print2DArr(arr)                         \
    for (int i = 0; i < arr.size(); i++)        \
    {                                           \
        for (int j = 0; j < arr[0].size(); j++) \
        {                                       \
            cout << arr[i][j] << " ";           \
        }                                       \
        cout << endl;                           \
    }

using namespace std;

class Solution
{
public:
    vector<int> successfulPairs(vector<int> &spells, vector<int> &potions, long long success)
    {
        int n = spells.size();
        int m = potions.size();
        vector<int> ans(n);
        sort(potions.begin(), potions.end());
        for (int i = 0; i < n; i++)
        {
            long long reqPot = (success / (long long)spells[i]);
            if (success % (long long)spells[i])
                reqPot++;
            if(reqPot > potions[m-1]) continue;
            int totalPairs = potions.end() - lower_bound(potions.begin(), potions.end(), int(reqPot));
            ans[i] = totalPairs;
        }
        return ans;
    }
};
